azure-devops/create-new-agent-image.ps1

```diff
@@ -0,0 +1,297 @@
+$ProtoVMName = 'PROTOTYPE'
+$LiveVMPrefix = 'BUILD'
+$WindowsServerSku = '2019-Datacenter'
+
+$ProgressActivity = 'Creating Scale Set'
+$TotalProgress = 8
+$CurrentProgress = 1
+
+function Find-ResourceGroupName {
+  Param(
+    [string] $prefix
+  )
+
+  $result = $prefix
+  $resources = Get-AzResourceGroup
+  $collision = $false
+  foreach ($resource in $resources) {
+    if ($resource.ResourceGroupName -eq $result) {
+      $collision = $true
+      break
+    }
+  }
+
+  $suffix = 0
+  while ($collision) {
+    $collision = $false
+    $suffix++
+    $result = "$prefix-$suffix"
+    foreach ($resource in $resources) {
+      if ($resource.ResourceGroupName -eq $result) {
+        $collision = $true
+        break
+      }
+    }
+  }
+
+  return $result
+}
+
+function New-Password {
+  Param (
+    [int] $length = 32
+  )
+
+  $Chars = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789"
+
+  $result = ''
+  for ($idx = 0; $idx -lt $length; $idx++) {
+    $result += $Chars[(Get-Random -Minimum 0 -Maximum ($Chars.Length - 1))]
+  }
+
+  return $result
+}
+
+function Start-WaitForShutdown {
+  Param([string]$ResourceGroupName, [string]$Name)
+  Write-Output "Waiting for $Name to stop..."
+  while ($true) {
+    $Vm = Get-AzVM -ResourceGroupName $ResourceGroupName -Name $Name -Status
+    $highestStatus = $Vm.Statuses.Count
+    for ($idx = 0; $idx -lt $highestStatus; $idx++) {
+      if ($Vm.Statuses[$idx].Code -eq 'PowerState/stopped') {
+        return
+      }
+    }
+
+    Write-Output "... not stopped yet, sleeping for 10 seconds"
+    Start-Sleep -Seconds 10
+  }
+}
+
+function Write-Reminders {
+  Param([string]$AdminPW)
+  Write-Output "Location: $Location"
+  Write-Output "Resource group name: $ResourceGroupName"
+  Write-Output "User name: AdminUser"
+  Write-Output "Using Generated Password: $AdminPW"
+}
+
+####################################################################################################
+Write-Progress `
+  -Activity $ProgressActivity `
+  -Status 'Creating resource group' `
+  -PercentComplete (100 / $TotalProgress * $CurrentProgress++)
+
+$ResourceGroupName = Find-ResourceGroupName $Prefix
+$AdminPW = New-Password
+Write-Reminders $AdminPW
+New-AzResourceGroup -Name $ResourceGroupName -Location $Location
+$AdminPWSecure = ConvertTo-SecureString $AdminPW -AsPlainText -Force
+$Credential = New-Object System.Management.Automation.PSCredential ("AdminUser", $AdminPWSecure)
+
+####################################################################################################
+Write-Progress `
+  -Activity 'Creating prototype VM' `
+  -PercentComplete (100 / $TotalProgress * $CurrentProgress++)
+
+$allowHttp = New-AzNetworkSecurityRuleConfig `
+  -Name AllowHTTP `
+  -Description 'Allow HTTP(s)' `
+  -Access Allow `
+  -Protocol Tcp `
+  -Direction Outbound `
+  -Priority 1008 `
+  -SourceAddressPrefix * `
+  -SourcePortRange * `
+  -DestinationAddressPrefix * `
+  -DestinationPortRange @(80, 443)

```

Comment:
Do we actually need to allow plain HTTP?

Reply:
I *think* we do but I wouldn't be surprised if not. In this case I'd prefer to keep it. The purpose of these firewall rules is primarily to prevent someone from turning our VMs into email spam bots so being a bit more open than necessary if it ever makes diagnosis of anything easier is preferrable.

Reply:
Ok.